coremltools/converters/mil/frontend/_utils.py

```diff
@@ -39,6 +41,36 @@ def value_at(x: Var, idx: int, name=None, before_op=None):
     return mb.slice_by_index(**args)
 
 
+def _construct_gather_op(op_type: str, x: Var, indices: Var, axis: Var = None, **kwarg) -> Var:
+    """
+    This utility is a more general gather in the sense that:
+    1. Both mb.gather and mb.gather_nd are handled
+    2. x is allowed to be bool, while mb.gather and mb.gather_nd only allow float or int
+    """
+    assert (
+        op_type in {"gather", "gather_nd"}
+    ), f"This utility only handles gather or gather_nd, but got {op_type}"
+    assert (
+        (op_type == "gather_nd") != (axis is not None)
+    ), "mb.gather_nd should not have input axis"
+
+    if x.dtype == types.bool:
```

Comment:
Will it be cleaner to do:
```
x_dtype = x.dtype

if x_dtype == types.bool:
    x = mb.cast(...)

if op_type == "gather":
   x = gather(..., name=name)
else:
  assert op_type == 'gather_nd"
  x = gather_nd(..., name=name)

if x_dtype == types.bool:
    x = mb.cast(..., name=name)
```
I think this would enhance readibility

Reply:
Note that I think the converter deduplicated the var name if there are two vars having the same name.
I am not sure about this, if it does not, we can always handle that ourselves too.

Reply:
Lemme give it a try 🚀 Indeed I thought I can only have name specified in the last `mb` op

Reply:
How handling by
```
is_gathering_bool = x.dtype == types.bool
    if is_gathering_bool:
        gather_name_kwarg = {}
        cast_name_kwarg = {} if name is None else {"name": name}
    else:
        gather_name_kwarg = {} if name is None else {"name": name}

    if is_gathering_bool:
        work_dtype = "int8" if is_current_opset_version_compatible_with(target.iOS17) else "fp16"
        x = mb.cast(x=x, dtype=work_dtype)

    if op_type == "gather":
        result = mb.gather(x=x, indices=indices, axis=axis, **gather_name_kwarg)
    else:
        result = mb.gather_nd(x=x, indices=indices, **gather_name_kwarg)

    if is_gathering_bool:
        result = mb.cast(x=result, dtype="bool", **cast_name_kwarg)
```